arrows-stdlib-cats/src/main/scala/arrows/stdlib/cats.scala

```diff
@@ -0,0 +1,114 @@
+package arrows.stdlib
+
+import cats.{ Applicative, StackSafeMonad }
+import cats.effect._
+import cats.arrow.ArrowChoice
+import cats.kernel.{ Monoid, Semigroup }
+import cats.mtl.ApplicativeAsk
+
+import scala.concurrent.{ ExecutionContext, Promise }
+import scala.util.control.NonFatal
+import scala.util.{ Failure, Success }
+
+object Cats extends ArrowInstances
+
+sealed abstract class ArrowInstances extends ArrowInstances1 {
+
+  implicit def catsEffectForTask(implicit ec: ExecutionContext): Effect[Task] = new ArrowAsync[Unit] with Effect[Task] {
+    def runAsync[A](fa: Task[A])(cb: Either[Throwable, A] => IO[Unit]): SyncIO[Unit] =
+      SyncIO { fa.run(())(ec); () }
+  }
+
+  implicit val catsArrowChoiceForArrow: ArrowChoice[Arrow] = new ArrowChoice[Arrow] {
+    def choose[A, B, C, D](f: Arrow[A, C])(g: Arrow[B, D]): Arrow[Either[A, B], Either[C, D]] =
+      Arrow[Either[A, B]].flatMap {
+        case Left(a)  => f(a).map(Left(_))
+        case Right(b) => g(b).map(Right(_))
+      }
+
+    def lift[A, B](f: A => B): Arrow[A, B] = Arrow[A].map(f)
+
+    def first[A, B, C](fa: Arrow[A, B]): Arrow[(A, C), (B, C)] =
+      Arrow[(A, C)].flatMap { case (a, c) => fa(a).map(_ -> c) }
+
+    def compose[A, B, C](f: Arrow[B, C], g: Arrow[A, B]): Arrow[A, C] = g.andThen(f)
+  }
+
+  implicit def catsApplicativeAskForArrow[E]: ApplicativeAsk[Arrow[E, ?], E] = new ApplicativeAsk[Arrow[E, ?], E] {
+    val applicative: Applicative[Arrow[E, ?]] = catsAsyncForArrow[E]
+
+    def ask: Arrow[E, E] = Arrow[E]
+
+    def reader[A](f: E => A): Arrow[E, A] = ask.map(f)
+
+  }
+
+  implicit def catsMonoidForArrow[A, B](implicit B0: Monoid[B]): Monoid[Arrow[A, B]] =
+    new Monoid[Arrow[A, B]] with ArrowSemigroup[A, B] {
+      implicit def B: Semigroup[B] = B0
+
+      def empty: Arrow[A, B] = Arrow.successful(Monoid[B].empty)
+    }
+}
+
+sealed abstract class ArrowInstances1 {
+
+  implicit def catsAsyncForArrow[E]: Async[Arrow[E, ?]] = new ArrowAsync[E] {}
+
+  implicit def catsSemigroupForArrow[A, B](implicit B0: Semigroup[B]): Semigroup[Arrow[A, B]] = new ArrowSemigroup[A, B] {
+    implicit def B: Semigroup[B] = B0
+  }
+}
+
+trait ArrowAsync[E] extends StackSafeMonad[Arrow[E, ?]] with Async[Arrow[E, ?]] {
+  def flatMap[A, B](fa: Arrow[E, A])(f: A => Arrow[E, B]): Arrow[E, B] =
+    Arrow[E].flatMap(e => fa(e).flatMap(a => f(a)(e)))
+
+  def raiseError[A](e: Throwable): Arrow[E, A] = Arrow.failed[A](e)
+
+  def handleErrorWith[A](fa: Arrow[E, A])(f: Throwable => Arrow[E, A]): Arrow[E, A] =
+    Arrow[E].flatMap(e => fa.recoverWith { case t: Throwable => f(t)(e) }(e))
+
+  def pure[A](x: A): Arrow[E, A] = Arrow.successful(x)
+
+  def suspend[A](thunk: => Arrow[E, A]): Arrow[E, A] =
+    Arrow[E].flatMap(e => try { thunk(e) } catch { case NonFatal(t) => Arrow.failed[A](t)(e) })
```

Comment:
there's no need to catch here

Reply:
Are you sure? If I change this line, it evaluates the line eagerly:

```
scala> def y(): Arrow[Int, String] = { throw new Exception(); }
y: ()arrows.stdlib.Arrow[Int,String]

scala> Arrow[Int].flatMap(y())
java.lang.Exception
  at .y(<console>:14)
  ... 36 elided
```

If I use my implementation instead:

```
scala> Arrow[Int].flatMap(i => try { y()(i) } catch { case t => Arrow.failed[String](t)(i) })
res6: arrows.stdlib.Arrow[Int,String] = <function1>
```

`Sync#suspend` requires to suspend any side-effects that occur when acquring the `Arrow`.

Reply:
yeah, you can't pass the arrow directly since it'll evaluate the by-name param. This should do it:

```scala
Arrow[E].flatMap(e => thunk(e))
```